bundle/edu.gemini.ags/src/main/scala/edu/gemini/ags/gems/mascot/Mascot.scala

```diff
@@ -77,23 +81,29 @@ object Mascot {
    * Finds the best asterisms for the given list of stars.
    * @param starList unfiltered list of stars from a catalog query
    * @param factor multiply strehl min, max and average by this value (depends on instrument filter: See REL-426)
-   * @param filter a filter function that returns false if the Star should be excluded
    * @param progress a function(strehl, count, total) called for each asterism as it is calculated
+   * @param filter a filter function that returns false if the Star should be excluded
+   * @param asterismPreFilter a filter function that returns false if a candidate asterism should be excluded
    * @return a tuple: (list of stars actually used, list of asterisms found)
    */
   def findBestAsterism(starList: List[Star],
                        factor: Double = defaultFactor,
                        progress: ProgressFunction = defaultProgress,
-                       filter: Star => Boolean = defaultFilter)
+                       filter: Star => Boolean = defaultFilter,
+                       asterismPreFilter: List[SiderealTarget] => Boolean = defaultAsterismPreFilter)
   : (List[Star], List[Strehl]) = {
     // sort by selected mag and select
     val sortedStarList = starList.sortWith((s1,s2) => s1.r < s2.r)
     val filteredStarList = selectStarsOnMag(sortedStarList).filter(filter)
-
     val ns = filteredStarList.length
-    val trips = AllPairsAndTriples.allTrips(filteredStarList)
-    val pairs = AllPairsAndTriples.allPairs(filteredStarList)
-    val total = trips.length + pairs.length + ns
+
+    // Create all candidate asterisms, filtering out those that are inadmissible.
+    val sings = filteredStarList.filter(s => asterismPreFilter(List(s.target)))
+    val trips = AllPairsAndTriples.allTrips(filteredStarList).
+      filter { case (s1,os2, os3) => asterismPreFilter(List(s1.target.some, os2.map(_.target), os3.map(_.target)).flatten) }
```

Comment:
If it is `allTrips` (i.e., all triples) why are stars 2, and 3 wrapped in an `Option`?  I would think a triple where one or more of the three is `None` would stop being a triple.  Same for `allPairs` below.

Reply:
I was indeed wondering the exact same thing when I came across this code.
I'm really not sure why a triple is a `(Star, Option[Star], Option[Star])` and a pair is a `(Star, Option[Star])'`.

I was hoping that maybe you knew? 😸